File: backend/internal/handler/playback.go
package handler

import (
	"github.com/thawng/velox/internal/auth"
	"github.com/thawng/velox/internal/model"
	"github.com/thawng/velox/internal/service"
	"github.com/thawng/velox/internal/transcoder"
)

var supportsSubtitleBurnIn = transcoder.SupportsSubtitleBurnIn

// PlaybackHandler provides playback decision and info endpoints
type PlaybackHandler struct {
	mediaSvc      *service.MediaService
	streamSvc     *service.StreamService
	userDataSvc   *service.UserDataService
	subtitleSvc   *service.SubtitleService
	audioTrackSvc *service.AudioTrackService
	markerSvc     *service.MarkerService
	configSvc     *service.PlaybackConfigService
	apiKeyStore   *auth.APIKeyStore
	activitySvc   *service.ActivityService
}

// NewPlaybackHandler creates a new playback handler
func NewPlaybackHandler(
	mediaSvc *service.MediaService,
	streamSvc *service.StreamService,
	userDataSvc *service.UserDataService,
	subtitleSvc *service.SubtitleService,
	audioTrackSvc *service.AudioTrackService,
	markerSvc *service.MarkerService,
	configSvc *service.PlaybackConfigService,
	apiKeyStore *auth.APIKeyStore,
) *PlaybackHandler {
	return &PlaybackHandler{
		mediaSvc:      mediaSvc,
		streamSvc:     streamSvc,
		userDataSvc:   userDataSvc,
		subtitleSvc:   subtitleSvc,
		audioTrackSvc: audioTrackSvc,
		markerSvc:     markerSvc,
		configSvc:     configSvc,
		apiKeyStore:   apiKeyStore,
	}
}

func (h *PlaybackHandler) SetActivityService(svc *service.ActivityService) {
	h.activitySvc = svc
}

// PlaybackInfoRequest represents client-sent capabilities
type PlaybackInfoRequest struct {
	VideoCodecs        []string `json:"video_codecs,omitempty"`
	AudioCodecs        []string `json:"audio_codecs,omitempty"`
	Containers         []string `json:"containers,omitempty"`
	MaxHeight          int      `json:"max_height,omitempty"`
	PreferDirectPlay   bool     `json:"prefer_direct_play,omitempty"`
	MediaFileID        int64    `json:"media_file_id,omitempty"`
	SelectedAudioTrack int      `json:"selected_audio_track,omitempty"`
	SelectedSubtitle   string   `json:"selected_subtitle,omitempty"`
	SelectedSubtitleID int      `json:"selected_subtitle_id,omitempty"`
}

// PlaybackInfoResponse represents playback decision response
type PlaybackInfoResponse struct {
	MediaID            int                 `json:"media_id"`
	PrimaryFileID      int64               `json:"primary_file_id,omitempty"`
	StreamSessionID    string              `json:"stream_session_id,omitempty"`
	Method             string              `json:"method"`
	StreamURL          string              `json:"stream_url"`
	DirectURL          string              `json:"direct_url,omitempty"`
	AbrURL             string              `json:"abr_url,omitempty"`
	VideoCodec         string              `json:"video_codec"`
	VideoProfile       string              `json:"video_profile,omitempty"`
	VideoLevel         int                 `json:"video_level,omitempty"`
	VideoFPS           float64             `json:"video_fps,omitempty"`
	AudioCodec         string              `json:"audio_codec"`
	Container          string              `json:"container"`
	FileSize           int64               `json:"file_size,omitempty"`
	Bitrate            int                 `json:"bitrate,omitempty"`
	Duration           int                 `json:"duration,omitempty"`
	Width              int                 `json:"width,omitempty"`
	Height             int                 `json:"height,omitempty"`
	AudioTracks        []AudioTrackInfo    `json:"audio_tracks,omitempty"`
	SubtitleTracks     []SubtitleTrackInfo `json:"subtitle_tracks,omitempty"`
	DecisionReason     string              `json:"decision_reason"`
	EstimatedBitrate   int                 `json:"estimated_bitrate,omitempty"`
	PtVideoCodec       string              `json:"pt_video_codec,omitempty"`
	PtAudioCodec       string              `json:"pt_audio_codec,omitempty"`
	PtHeight           int                 `json:"pt_height,omitempty"`
	PtVideoBitrate     int                 `json:"pt_video_bitrate,omitempty"`
	PtAudioBitrate     int                 `json:"pt_audio_bitrate,omitempty"`
	Position           float64             `json:"position,omitempty"`
	SkipSegments       []model.SkipSegment `json:"skip_segments,omitempty"`
	AvailableQualities []QualityOption     `json:"available_qualities,omitempty"`
}

// AudioTrackInfo represents an audio track
type AudioTrackInfo struct {
	ID         int    `json:"id"`
	Language   string `json:"language"`
	Label      string `json:"label"`
	Codec      string `json:"codec,omitempty"`
	Channels   int    `json:"channels,omitempty"`
	Bitrate    int    `json:"bitrate,omitempty"`
	SampleRate int    `json:"sample_rate,omitempty"`
	IsDefault  bool   `json:"is_default"`
	Selected   bool   `json:"selected"`
}

// SubtitleTrackInfo represents a subtitle track
type SubtitleTrackInfo struct {
	ID        int    `json:"id"`
	Language  string `json:"language"`
	Label     string `json:"label"`
	Format    string `json:"format"`
	IsDefault bool   `json:"is_default"`
	IsImage   bool   `json:"is_image"`
}

File: packages/shared/types/playback.ts
// Playback, Streaming, Subtitle types
import type { ImageResource } from './common'

export interface StreamUrls {
  /** Original file with pm=direct (forces backend to bypass pretranscode lookup). */
  direct: string
  /** Pre-encoded MP4 endpoint (no pm), only set when a ready pre-transcode file exists. */
  pretranscode?: string
  /** Realtime HLS transcode endpoint, always present. */
  hls: string
  /** Backend hint: which URL to try first. Defaults to 'direct'. */
  prefer?: 'direct' | 'pretranscode' | 'hls'
  primary_file_id: number
  stream_session_id: string
}

export interface StreamURLResponse {
  direct_url: string
  hls_url: string
  stream_session_id: string
  api_key: string
  expires_in: number
}

export interface SubtitleTrack {
  id: number
  media_id: number
  language: string
  label: string
  file_path: string
  is_default: boolean
}

export interface AudioTrack {
  id: number
  media_id: number
  language: string
  label: string
  codec: string
  channels: number
  is_default: boolean
}

export interface PlaybackInfoRequest {
  video_codecs?: string[]
  audio_codecs?: string[]
  containers?: string[]
  max_height?: number
  media_file_id?: number
  selected_audio_track?: number
  selected_subtitle?: string
  selected_subtitle_id?: number
}

export interface SkipSegment {
  type: 'intro' | 'credits'
  start: number
  end: number
  source: string
  confidence: number
}

export interface PlaybackSubtitleTrack {
  id: number
  language: string
  label: string
  format: string
  is_default: boolean
  is_image: boolean
}

export interface PlaybackAudioTrack {
  id: number
  language: string
  label: string
  codec: string
  channels: number
  bitrate: number
  sample_rate: number
  is_default: boolean
  selected: boolean
}

export interface SubtitleSearchResult {
  provider: 'opensubtitles' | 'subdl'
  external_id: string
  title: string
  language: string
  format: string
  downloads: number
  rating: number
  forced: boolean
  hearing_impaired: boolean
  ai_translated: boolean
}

export interface SubtitleDownloadRequest {
  provider: string
  external_id: string
  language: string
}

export type PlaybackPrefer = 'direct' | 'pretranscode' | 'hls'

export interface PlaybackInfo {
  media_id: number
  primary_file_id: number
  stream_session_id: string
  method: string
  stream_url: string
  direct_url: string
  /** Same media_id endpoint without pm=, so backend's pretranscode lookup runs.
   *  Only present when a ready pre-transcode file exists. */
  pretranscode_url?: string
  /** Realtime HLS transcode endpoint, always present as last-resort fallback. */
  hls_url: string
  /** Backend hint of which URL to try first. Client owns the actual fallback chain. */
  prefer?: PlaybackPrefer
  abr_url?: string
  video_codec: string
  video_profile: string
  video_level: number
  video_fps: number
  audio_codec: string
  container: string
  file_size: number
  bitrate: number
  duration: number
  width: number
  height: number
  audio_tracks: PlaybackAudioTrack[]
  subtitle_tracks: PlaybackSubtitleTrack[]
  decision_reason: string
  estimated_bitrate: number
  pt_video_codec?: string
  pt_audio_codec?: string
  pt_height?: number
  pt_video_bitrate?: number
  pt_audio_bitrate?: number
  position: number
  skip_segments?: SkipSegment[]
  available_qualities?: QualityOption[]
}

export interface QualityOption {
  height: number
  label: string
  instant: boolean
  source: 'original' | 'pretranscode' | 'transcode'
  bitrate_kbps?: number
}

// User progress & continue watching

export interface UserData {
  user_id: number
  media_id: number
  position: number
  completed: boolean
  is_favorite: boolean
  rating?: number
  play_count: number
  last_played_at?: string
  updated_at: string
  // JOIN fields
  media_title?: string
  poster?: ImageResource | null
  media_duration?: number
}

export interface UpdateProgressRequest {
  position: number
  completed: boolean
}

export interface ToggleFavoriteResponse {
  is_favorite: boolean
}

export interface FavoritesListParams {
  limit?: number
  offset?: number
  start_char?: string
}

export interface RecentlyWatchedParams {
  limit?: number
}

export interface ContinueWatchingItem {
  media_id: number
  series_id?: number
  position: number
  completed: boolean
  last_played_at?: string
  title: string
  poster?: ImageResource | null
  backdrop?: ImageResource | null
  media_type: 'movie' | 'episode'
  duration: number
  series_title?: string
  season_number?: number
  episode_number?: number
}

export interface NextUpItem {
  media_id: number
  series_id: number
  title: string
  episode_title: string
  media_type: 'episode'
  poster?: ImageResource | null
  backdrop?: ImageResource | null
  still?: ImageResource | null
  duration: number
  season_number: number
  episode_number: number
  series_title: string
  last_watched_at?: string
}

export interface ContinueWatchingParams {
  limit?: number
}

export interface NextUpParams {
  limit?: number
}
